lms/services/vitalsource/_client.py

```diff
@@ -76,12 +80,112 @@ def get_table_of_contents(self, book_id: str):
 
         return toc
 
-    def _json_request(self, method, endpoint):
+    def get_user_book_license(self, user_reference, book_id):
+        """
+        Get a user licence for a specific book (if any).
+
+        :param user_reference: String identifying the current user
+        :param book_id: The book to get the licence for
+        """
+        result = self._xml_request(
+            "GET",
+            f"{self.VS_API}/v3/licenses.xml",
+            params={"sku": book_id},
+            auth=self._user_auth(user_reference),
+        )
+
+        # The result is a list of active licenses that match the given book ID/SKU.
+        try:
+            return result["licences"]["license"]
+        except KeyError:
+            return None
+
+    def get_sso_redirect(self, user_reference, url):
+        """
+        Get a URL which logs in a user then redirects to a URL.
+
+        :param user_reference: String identifying the current user
+        :param url: The URL to redirect to after login
+        """
+        result = self._xml_request(
+            "POST",
+            f"{self.VS_API}/v3/redirects.xml",
+            data={"redirect": {"destination": url}},
+            auth=self._user_auth(user_reference),
+        )
+        return result["redirect"]["@auto-signin"]
+
+    def _user_auth(self, user_reference):
+        # Implement requests auth interface, which is just a callable which
+        # gets the request object and modifies it.
+        def _user_auth(request):
+            request.headers["X-VitalSource-Access-Token"] = self._get_user_access_token(
+                user_reference
+            )
+
+            return request
+
+        return _user_auth
```

Comment:
This method looks a bit strange, but has a nice effect for us, in that it doesn't involve having to merge headers anywhere.

The way this is setup it also means you can call any method in any order without having to worry about auth for yourself, and we will only ever make one call to get the access token.